Test for reading JDK 8 (classfile format 52) class files.

JamesIry opened the pull request:
This commit includes a test for reading JDK 8 (classfile format 52)
class files, in particular default (aka defender) methods. It uses ASM
to generate an interface with default methods then exercises that
interface from Scala.

Surprisingly no changes are necessary to the Scala code base to support
reading format 52 class files.

Because the test can only run under JDK 8, the JDK version is checked
and the expected output is synthesized for previous versions.

Commit:
Test for reading JDK 8 (classfile format 52) class files.

This commit includes a test for reading JDK 8 (classfile format 52)
class files, in particular default (aka defender) methods. It uses ASM
to generate an interface with default methods then exercises that
interface from Scala.

Surprisingly no changes are necessary to the Scala code base to support
reading format 52 class files.

Because the test can only run under JDK 8, the JDK version is checked
and the expected output is synthesized for previous versions.

JamesIry commented:
review @gkossakowski  

Note: I've run a hacked version of the code that throws an error on JDK < 8 to verify that it's actually testing something. To get JDK visit https://jdk8.java.net/download.html

retronym commented:
The JDK version checking is done a little differently here: b2c67b328. Might pop this into a utility method in partest.

JamesIry commented:
Ah, @paulp  did the Java 8 classfile work https://github.com/scala/scala/pull/2456. So let me rephrase, "surprisingly it worked in spite of not having a test for class file reading." So there!

JamesIry commented:
@retronym good idea. And I'll update https://github.com/scala/scala/blob/4c85fb324dc3ea0eeebe590e39e8084558834f3a/test/files/run/classfile-format-51.scala

soc commented on test/files/run/classfile-format-52.scala:
```diff
@@ -0,0 +1,80 @@
+      // cheat a little by using 'ScalaVersion' because it can parse java versions just as well
+      val requiredJavaVersion = ScalaVersion("1.8")
+      val executingJavaVersion = ScalaVersion(System.getProperty("java.specification.version"))
+      if (executingJavaVersion >= requiredJavaVersion) {
```
`isJavaAtLeast("1.8")`?

Commit:
Refactor testing logic for only running under certain JDK versions

We had several tests designed to only run if the JDK version was at
least some specified version. This commit refactors that common logic
into DirectTest.

retronym commented:
LGTM. Seeing as it's just tests I'm going to merge, but I'd encourage @gkossakowski to still take a look at this after the fact.

gkossakowski commented on src/partest/scala/tools/partest/DirectTest.scala:
```diff
@@ -101,4 +102,30 @@ abstract class DirectTest extends App {
+  def testUnderJavaAtLeast[A](version: String)(yesRun: =>A) = new TestUnderJavaAtLeast(version, { yesRun })
+
+  class TestUnderJavaAtLeast[A](version: String, yesRun: => A) {
+    val javaVersion = System.getProperty("java.specification.version")
```
Interesting. I didn't know there's property like that.

JamesIry replied:
Yeah, much easier to parse than the full version info.

paulp replied:
What the... that property needs a better PR agent!

gkossakowski commented:
LGTM.

Sorry for the delay, I didn't realize it was sitting in my queue. I'll pay more attention to it now.
